pkg/ingester/instance.go

```diff
@@ -215,13 +211,10 @@ func (i *instance) Query(req *logproto.QueryRequest, queryServer logproto.Querie
 		},
 	)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	iter := iter.NewHeapIterator(ctx, iters, req.Direction)
-	defer helpers.LogError("closing iterator", iter.Close)
-
-	return sendBatches(ctx, iter, queryServer, req.Limit)
+	return iter.NewHeapIterator(ctx, iters, req.Direction), nil
```

Comment:
To avoid multiple HeapIterator nesting, which can slow down execution, you could return an array of iterators here and apprends the store later if needed.